src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,712 @@
+                {
+                    _dictionaryStorage = new Dictionary<TKey, TValue>(capacity);
+                    foreach (var item in _arrayStorage)
+                    {
+                        _dictionaryStorage[item.Key] = item.Value;
+                    }
+
+                    // Clear array storage.
+                    _arrayStorage = null;
+                }
+                else
+                {
+                    capacity = _arrayStorage.Length == 0 ? DefaultArrayThreshold : _arrayStorage.Length * 2;
+                    var array = new KeyValuePair<TKey, TValue>[capacity];
+                    if (_count > 0)
+                    {
+                        Array.Copy(_arrayStorage, 0, array, 0, _count);
+                    }
+
+                    _arrayStorage = array;
+                }
+            }
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int FindIndex(TKey key)
+        {
+            // Generally the bounds checking here will be elided by the JIT because this will be called
+            // on the same code path as EnsureCapacity.
+            Debug.Assert(_dictionaryStorage == null);
+            Debug.Assert(_arrayStorage != null);
+
+            var array = _arrayStorage;
+            var count = _count;
+
+            for (var i = 0; i < count; i++)
+            {
+                if (_comparer.Equals(array[i].Key, key))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryFindItem(TKey key, out TValue? value)
+        {
+            Debug.Assert(_dictionaryStorage == null);
+            Debug.Assert(_arrayStorage != null);
+
+            var array = _arrayStorage;
+            var count = _count;
+
+            if ((uint)count <= (uint)array.Length)
+            {
+                for (var i = 0; i < count; i++)
+                {
+                    if (_comparer.Equals(array[i].Key, key))
+                    {
+                        value = array[i].Value;
+                        return true;
+                    }
+                }
+            }
+
+            value = default;
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool ContainsKeyArray(TKey key)
+        {
+            Debug.Assert(_dictionaryStorage == null);
+            Debug.Assert(_arrayStorage != null);
+
+            var array = _arrayStorage;
+            var count = _count;
+
+            if ((uint)count <= (uint)array.Length)
+            {
+                for (var i = 0; i < count; i++)
+                {
+                    if (_comparer.Equals(array[i].Key, key))
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
```

Comment:
```suggestion
        private bool ContainsKeyArray(TKey key) => TryFindItem(key, out _);
```
I verifyed the codegen and perf. Looks good that way, as the JIT eliminates the dead unnecessary code due to inlining.